src/interpreter.ts

```diff
@@ -485,18 +491,46 @@ export class Interpreter {
     private currentEnv: RuntimeEnvironment;
     private initialContext: CompilerContext;
     private config: InterpreterConfig;
-    private globalEnv: RuntimeEnvironment;
 
     constructor(
         initialContext: CompilerContext = new CompilerContext(),
         config: InterpreterConfig = defaultInterpreterConfig,
     ) {
-        this.globalEnv = new RuntimeEnvironment();
-        this.currentEnv = this.globalEnv;
+        this.currentEnv = new RuntimeEnvironment();
         this.initialContext = initialContext;
         this.config = config;
     }
 
+    // Opens a new environment with the provided param names and
+    // values.
+    private openNewEnvironment(
+        args: { paramNames: string[]; values: Value[] } = {
+            paramNames: [],
+            values: [],
+        },
+    ) {
+        const paramNames = args.paramNames;
+        const values = args.values;
+
+        this.currentEnv = new RuntimeEnvironment(this.currentEnv);
+        paramNames.forEach((param, index) => {
+            this.currentEnv.setNewBinding(param, values[index]!);
+        }, this);
+    }
+
+    // Closes the current environment and returns to its
+    // parent environment
+    private closeCurrentEnvironment() {
+        const parentEnv = this.currentEnv.getParentEnvironment();
+        if (parentEnv === undefined) {
+            // This is a programmer's error, not even a compilation error.
+            throw new Error(
```

Comment:
See PR #669 which uses "internal compiler error" instead of throwing just an `Error`